content/number-theory/MillerRabin.h

```diff
@@ -1,30 +1,27 @@
 /**
- * Author: Lukas Polacek
- * Date: 2010-01-26
+ * Author: chilli, c1729
+ * Date: 2019-03-28
  * License: CC0
- * Source: TopCoder tutorial
- * Description: Miller-Rabin primality probabilistic test.
- * Probability of failing one iteration is at most 1/4. 15 iterations should be
- * enough for 50-bit numbers.
- * Time: 15 times the complexity of $a^b \mod c$.
+ * Description: Deterministic Miller-Rabin primality test.
+ * Guaranteed to be correct, 2x as probabilistic version for <1e9, 3x as fast
+ * for 1e15
+ * Time: 7 times the complexity of $a^b \mod c$.
  */
 #pragma once
 
 #include "ModMulLL.h"
 
-bool prime(ull p) {
-	if (p == 2) return true;
-	if (p == 1 || p % 2 == 0) return false;
-	ull s = p - 1;
-	while (s % 2 == 0) s /= 2;
-	rep(i,0,15) {
-		ull a = rand() % (p - 1) + 1, tmp = s;
-		ull mod = mod_pow(a, tmp, p);
-		while (tmp != p - 1 && mod != 1 && mod != p - 1) {
-			mod = mod_mul(mod, mod, p);
-			tmp *= 2;
-		}
-		if (mod != p - 1 && tmp % 2 == 0) return false;
-	}
-	return true;
+bool prime(ull n) {
+    vector<ull> ps({2, 3, 5, 13, 19, 73, 193, 407521, 299210837});
+    vector<ull> cs({2, 325, 9375, 28178, 450775, 9780504, 1795265022});
```

Comment:
Can you add a reference for these numbers? For which range of `n` do they suffice? I can imagine wanting to reimplement this in Python with bignums during a contest, so 2**64 isn't always enough and we should describe how to safely extend it to larger n (is "test 15 a's at random" the right method for that?).

Reply:
Well, it was originally some Python code I made, from [here](https://github.com/cheran-senthil/PyRival/blob/master/snippets/Mathematics/is_prime.py), perhaps you can come up with a cleaner convertion to C++. You can safely extend to more numbers if you are checking for over 2**64, by taking [more smaller primes](https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test#Deterministic_variants) upto 2(ln n)2

Reply:
I think if you want to change it to a randomized version that'll "work" for arbitrarily large numbers, you can just select 15 random a's.

You shouldn't need to change anything else.

Reply:
Also, to be clear, the numbers are taken from https://miller-rabin.appspot.com/

Reply:
Also, where should I add this reference? Seems unnecessary to put the source of the numbers in the notebook itself, but I'm not sure what's appropriate. Perhaps the source/authors?

Reply:
Source would be good.